engine/api/workflow/hook.go

```diff
@@ -0,0 +1,178 @@
+package workflow
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/engine/api/cache"
+	"github.com/ovh/cds/engine/api/repositoriesmanager"
+	"github.com/ovh/cds/engine/api/services"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+)
+
+// HookRegistration
+func HookRegistration(db gorp.SqlExecutor, store cache.Store, oldW *sdk.Workflow, wf sdk.Workflow, p *sdk.Project) error {
+	var hookToUpdate map[string]sdk.WorkflowNodeHook
+	var hookToDelete map[string]sdk.WorkflowNodeHook
+
+	if oldW != nil {
+		hookToUpdate, hookToDelete = diffHook(oldW.GetHooks(), wf.GetHooks())
+	} else {
+		hookToUpdate = wf.GetHooks()
+	}
+
+	if len(hookToUpdate) > 0 {
+		//Push the hook to hooks µService
+		dao := services.Querier(db, store)
+		//Load service "hooks"
+		srvs, err := dao.FindByType("hooks")
+		if err != nil {
+			return sdk.WrapError(err, "HookRegistration> Unable to get services dao")
+		}
+
+		// Update in VCS
+		for i := range hookToUpdate {
+			h := hookToUpdate[i]
+			if oldW != nil && wf.Name != oldW.Name {
+				configValue := h.Config["workflow"]
+				configValue.Value = wf.Name
+				h.Config["workflow"] = configValue
+				hookToUpdate[i] = h
+			}
+		}
+
+		//Perform the request on one off the hooks service
+		if len(srvs) < 1 {
+			return sdk.WrapError(fmt.Errorf("HookRegistration> No hooks service available, please try again"), "Unable to get services dao")
+		}
+
+		var hooksUpdated map[string]sdk.WorkflowNodeHook
+		code, errHooks := services.DoJSONRequest(srvs, http.MethodPost, "/task/bulk", hookToUpdate, &hooksUpdated)
+		if errHooks != nil || code >= 400 {
+			return sdk.WrapError(errHooks, "HookRegistration> Unable to create hooks")
+		}
+
+		for i := range hooksUpdated {
+			h := hooksUpdated[i]
+			if h.Config["vcsServer"].Value != "" {
+				if err := createVCSConfiguration(db, store, p, &h); err != nil {
+					return sdk.WrapError(err, "HookRegistration> Cannot update vcs configuration")
+				}
+			}
+			if err := UpdateHook(db, &h); err != nil {
+				return sdk.WrapError(errHooks, "HookRegistration> Cannot update hook")
+			}
+		}
+	}
+
+	if len(hookToDelete) > 0 {
+		for _, h := range hookToDelete {
+			if err := deleteVCSConfiguration(db, store, p, h); err != nil {
+				return sdk.WrapError(err, "HookRegistration> Cannot delete vcs configuration for hook %s", h.UUID)
+			}
+		}
+
+		//Push the hook to hooks µService
+		dao := services.Querier(db, store)
+		//Load service "hooks"
+		srvs, err := dao.FindByType("hooks")
+		if err != nil {
+			return sdk.WrapError(err, "HookRegistration> Unable to get services dao")
+		}
+		code, errHooks := services.DoJSONRequest(srvs, http.MethodDelete, fmt.Sprintf("/task/bulk"), hookToDelete, nil)
+		if errHooks != nil || code >= 400 {
+			log.Warning("HookRegistration> Unable to delete old hooks")
```

Comment:
no wrap here ?

Add http code and errHooks in msg

Reply:
I do not want to rollback because of an error on repo manager size.